webnn/resources/utils.js

```diff
@@ -1,61 +1,312 @@
+  if (typeof expected === 'number') {
+    // for checking a scalar output by matmul 1D x 1D
+    expected = [expected];
+    actual = [actual];
+  }
+  if (metricType === 'ULP') {
+    assert_array_approx_equals_ulp(actual, expected, tolerance, operandType, description);
+  } else if (metricType === 'ATOL') {
+    assert_array_approx_equals(actual, expected, tolerance, description);
+  }
+};
+
+/**
+ * Check computed results with expected data.
+ * @param {String} name - An operation name
+ * @param {Object.<String, MLOperand>} namedOutputOperands
+ * @param {Object.<MLNamedArrayBufferViews>} outputs - The resources of required outputs
+ * @param {Object} resources - Resources used for building a graph
+ */
+const checkResults = (name, namedOutputOperands, outputs, resources) => {
+  const metricType = PrecisionMetrics[name] ? Object.keys(PrecisionMetrics[name])[0] : 'ULP';
+  const tolerance = getPrecisonTolerance(name, metricType, resources);
+  const expected = resources.expected;
+  const operandType = expected.type;
+  let outputData;
+  let expectedData;
+  if (Array.isArray(expected)) {
+    // check outputs by split or gru
+    for (let operandName in namedOutputOperands) {
+      outputData = outputs[operandName];
+      expectedData = getExpectedData(expected, operandName);
+      doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
+    }
+  } else {
+    const outputName = expected.name ? expected.name : 'output';
+    outputData = outputs[outputName];
+    expectedData = expected.data;
+    doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
+  }
+};
+
+/**
+ * Build a graph.
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {Function} buildFunc - A build function for an operation
+ * @param {String} name - An operation name
+ * @returns [namedOperands, inputs, outputs]
+ */
+const buildGraph = (builder, resources, buildFunc, name) => {
+  const outputOperand = buildFunc(builder, resources, name);
+  let inputs = {};
+  let namedOperands = {};
+  if (Array.isArray(resources.inputs)) {
+    // the inputs of concat() is a sequence
+    for (let subInput of resources.inputs) {
+      inputs[subInput.name] = new TypedArrayDict[subInput.type](subInput.data);
+    }
+  } else {
+    for (let inputName in resources.inputs) {
+      const subTestByName = resources.inputs[inputName];
+      inputs[inputName] = new TypedArrayDict[subTestByName.type](subTestByName.data);
+    }
+  }
+  let outputs = {};
+  if (Array.isArray(resources.expected)) {
+    // the outputs of split() or gru() is a sequence
+    for (let i = 0; i < resources.expected.length; i++) {
+      const subExpected = resources.expected[i];
+      outputs[subExpected.name] = new TypedArrayDict[subExpected.type](sizeOfShape(subExpected.shape));
+      namedOperands[subExpected.name] = outputOperand[i];
```

Comment:
It'd be better to let `buildFunc()` return `namedOperands`. The mapping of `outputOperand[i]` to `namedOperands[subExpected.name]` is unclear.